### ISSUE TITLE
MacOSXFsEventsDiffAwarenessTest is flaky

### ISSUE THREAD

**Issue by philwo** [bazelbuild/bazel#10776]
//src/test/java/com/google/devtools/build/lib/skyframe:SkyframeTests has been failing on CI recently due to MacOSXFsEventsDiffAwarenessTest being flaky. The current theory is that the test fails when the machine is under high load.

When we looked at the test initially, it seemed quite obvious why it would be flaky - it did some changes to the filesystem, waited 200ms and then checked that FSEvents reported them correctly. It's conceivable that this might not be the case under load.

@jmmv kindly fixed the test and added logic to retry up to 60 seconds, however the test is still flaky (although apparently less than before):

Log: https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/391e674d-e4dc-4d2d-9b2d-e12d7e4082da/src/test/java/com/google/devtools/build/lib/skyframe/SkyframeTests/attempt_1.log
CI job: https://buildkite.com/bazel/bazel-bazel/builds/11455

This is bad, because it might indicate that our `--watchfs` implementation on macOS sometimes doesn't work correctly and Bazel might miss changed files during incremental rebuilds. 🤔

I'll disable the test for now, but we should look into this.

**Comment by jmmv** [comment]
Did anything change at all on the CI machines? I find it quite strange that "heavy load" has become a problem only now and is triggering this problem so frequently when we never saw it in the past.

**Comment by philwo** [comment]
@jmmv It's very hard to say. We upgraded the production fleet to macOS Catalina on 2020-01-29 (but on testing CI already much earlier, I think since the beta), and the test worked fine until yesterday. I remember seeing the same failure occasionally during a presubmit of my CL in the testing org, but that was a completely unrelated change, then it went away a few days later and I was able to submit it.

Maybe it's a bug in macOS Catalina, but it's not clear what triggers it :|

**Comment by jmmv** [comment]
Alright, I think I know what happens. The test starts by doing:

```java
View view1 = underTest.getCurrentView(watchFsEnabledProvider);
```

which in turn calls `MacOSXFsEventsDiffAwareness#init` to initialize the monitor. But the monitor is started in a separate thread. There is no guarantee that by the end of init, fsevents is already listening for changes, so if the thread is delayed, we can modify the client before fsevents starts to listen -- and thus we lose the initial events and we don't see anything ever changing.

Trivial to expose this by adding a sleep right before the call to `MacOSXFsEventsDiffAwareness#run`.

**Comment by jmmv** [comment]
Haha, and of course not yet fixed. Flaked again.

**Comment by jmmv** [comment]
I spent a lot of time on the theory that this is caused by heavy machine load but could not trigger a single problem on various machines (including an iMac Pro as BazelCI uses) by running the test hundreds of times while, at the same time, running a very large build in the background. I even "reniced" `fseventsd` to 20 to try to make it slower, but that had no effect either.

So I ended up writing a test that creates files concurrently from a bunch of threads (100 threads, 50 files each). This is not "load" in the sense as I was looking for earlier given that this completes in milliseconds and `fseventsd` barely uses any CPU. But bingo. Our diff awareness code is almost always failing to reliably detect all notifications. Use 1 thread and 5000 files and things work though, so it's not necessarily the number of files to monitor that's the problem.

Digging further, and reading through the `FSEvents.h` header (which is much more completely documented than the online documentation for some reason)... I encountered `kFSEventStreamEventFlagMustScanSubDirs`, which tells the caller that they are responsible for looking for further changes as events may have dropped. And lo and behold, if we look into the flags we get in `FsEventsDiffAwarenessCallback` when the test fails, we are being told about this condition. Further, we are told `kFSEventStreamEventFlagUserDropped`, which means we aren't processing events fast enough and they drop because of us.

In other words, our usage of the `FSEvents` API is incorrect and we are missing changes when we can't keep up with the notifications. We must interpret these flags to know when events dropped and do something about them. I guess the easiest solution on this condition is to "give up" and assume "everything changed" in the response. Looking into it.

cc @haxorz

**Comment by philwo** [comment]
Our good old friend is back again:
<img width="262" alt="Screenshot 2020-05-04 at 13 32 47" src="https://user-images.githubusercontent.com/504652/80962098-a6f44700-8e0c-11ea-91bf-f8c310e086aa.png">

Always failing with this:
```
There was 1 failure:
1) testSimple(com.google.devtools.build.lib.skyframe.MacOSXFsEventsDiffAwarenessTest)
java.lang.AssertionError: Paths [b/c/d, a, b, a/b, a/b/c, b/c] not found as modified
	at com.google.devtools.build.lib.skyframe.MacOSXFsEventsDiffAwarenessTest.assertDiff(MacOSXFsEventsDiffAwarenessTest.java:144)
	at com.google.devtools.build.lib.skyframe.MacOSXFsEventsDiffAwarenessTest.testSimple(MacOSXFsEventsDiffAwarenessTest.java:161)
```

I'm disabling the test again. 😭

**Comment by philwo** [comment]
Here's a failure log of `testStress` (which, ironically, fails less often than `testSimple`):

https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/6ed4846d-5d6d-4978-a94f-4e91dea77d3e/src/test/java/com/google/devtools/build/lib/skyframe/SkyframeTests/attempt_6.log

**Comment by jmmv** [comment]
FTR, I just submitted b03765aa1944976ea22cb3f90ca81682be105f0b which corrects yet another problem in this test, but I cannot reproduce the flakiness even before this fix...

**Comment by jmmv** [comment]
To make things worse, `testStress` gets stuck if run with `--runs_per_test` over about 10 because we end up with too many concurrent threads opening files and, I think, running into process limits without good error control. But still cannot reproduce the flakiness issues.

**Comment by jmmv** [comment]
Hahaha, and it's flaky again 😭

**Comment by susinmotion** [comment]
I'm not quite sure if I've put this in the correct component, but moving it out of team-Apple since it doesn't quite belong there. Adding untriaged label so that it gets picked up by the Core team's triage process.

**Comment by janakdr** [comment]
@susinmotion can you say why this doesn't belong in team-Apple? The test looks to be doing a lot of local execution/Mac-specific stuff that -core has no expertise on.

**Comment by susinmotion** [comment]
I tagged it as Core since fs events is used as part of the core Bazel functionality. This issue is indeed specific to Mac, but it should be considered like any other platform that is supported by Bazel. Imagine we had this issue on Windows or Linux. Would it belong in the core component? If not, then I've miscategorized it.

For some more context, the team-Apple component is (just now!) deprecated, after many months of winding down. This reflects the community ownership of rules_apple, and leaves space for core Bazel issues related to Apple in team-Rules-CPP. The Apple team remains responsible just for issues relating to the interfacing of Macs with production services.

**Comment by janakdr** [comment]
After some offline discussion, moving this to team-Local-Exec, since Core, despite the name, has no domain knowledge here :/
